packages/application-generic/src/usecases/add-job/add-job.usecase.ts

```diff
@@ -54,22 +62,47 @@ export class AddJob {
 
     Logger.log(`Starting New Job ${job._id} of type: ${job.type}`, LOG_CONTEXT);
 
-    const digestAmount =
-      job.type === StepTypeEnum.DIGEST
-        ? await this.addDigestJob.execute(AddDigestJobCommand.create({ job }))
-        : undefined;
-
+    let digestAmount: number | undefined;
+    let digestCreationResult: DigestCreationResultEnum | undefined;
     if (job.type === StepTypeEnum.DIGEST) {
+      validateDigest(job);
+      digestAmount = this.calculateDelayService.calculateDelay({
+        stepMetadata: job.digest,
+        payload: job.payload,
+        overrides: job.overrides,
+      });
       Logger.debug(`Digest step amount is: ${digestAmount}`, LOG_CONTEXT);
-    }
 
-    if (job.type === StepTypeEnum.DIGEST && digestAmount === undefined) {
-      Logger.warn(
-        `Digest Amount does not exist on a digest job ${job._id}`,
-        LOG_CONTEXT
+      digestCreationResult = await this.mergeOrCreateDigestUsecase.execute(
+        MergeOrCreateDigestCommand.create({ job })
       );
 
-      return;
+      if (digestCreationResult === DigestCreationResultEnum.MERGED) {
+        Logger.log('Digest was merged, queueing next job', LOG_CONTEXT);
+
+        return;
+      }
+
+      if (digestCreationResult === DigestCreationResultEnum.SKIPPED) {
+        const nextJobToSchedule = await this.jobRepository.findOne({
+          _environmentId: command.environmentId,
+          _parentId: job._id,
+        });
+
+        if (!nextJobToSchedule) {
+          return;
+        }
+
+        await this.execute({
+          userId: job._userId,
+          environmentId: job._environmentId,
+          organizationId: command.organizationId,
+          jobId: nextJobToSchedule._id,
+          job: nextJobToSchedule,
+        });
+
+        return;
+      }
```

Comment:
Is this recursive iteration to try to jump all the extra time spent in different calculations in the flow to minimise the race conditions for jobs that are meant to be skipped because they are digested?
I find this bit of flow easy to miss conceptually and not sure how it would interact with `QueueNextJob` that is executed in `RunJob` and at the end of its execution calls to `AddJob`.

Reply:
Yes that is the reason. If I remember correctly me and @scopsy had a discussion about how it would interact with `QueueNextJob` and ended up with that it would be okay.